usr.sbin/devinfo/devinfo.c

```diff
@@ -116,50 +183,85 @@ print_device_rman_resources(struct devinfo_rman *rman, void *arg)
 	    print_device_matching_resource, ia) != 0) {
 
 		/* there are, print header */
-		for (i = 0; i < indent; i++)
-			printf(" ");
-		printf("%s:\n", rman->dm_desc);
+
+		safe_desc = (char*) malloc(sizeof(rman->dm_desc));
+		strcpy(safe_desc, rman->dm_desc);
+		xml_safe_string(safe_desc);
+
+		print_indent(indent);
+		xo_emit("{d:%s}:\n", rman->dm_desc);
+		xo_open_list(safe_desc);
 
 		/* print resources */
 		ia->indent = indent + 4;
 		devinfo_foreach_rman_resource(rman,
-		    print_device_matching_resource, ia);
+			print_device_matching_resource, ia);
+
+		xo_close_list(safe_desc);
 	}
 	ia->indent = indent;
 	return(0);
 }
 
 static void
-print_dev(struct devinfo_dev *dev)
+print_device_props(struct devinfo_dev *dev)
 {
+	if (vflag) {
+		if (*dev->dd_pnpinfo) {
+			xo_open_container("pnpinfo");
+			xo_emit("{D: pnpinfo}");
+
+			if ((strcmp(dev->dd_pnpinfo, "unknown") == 0))
+				xo_emit("{D: unknown}");
+			else
+				print_kvlist(dev->dd_pnpinfo);
+
+			xo_close_container("pnpinfo");
+		}
+		if (*dev->dd_location) {
+			xo_open_container("location");
+			xo_emit("{D: at}");
+			print_kvlist(dev->dd_location);
+			xo_close_container("location");
+		}
+
+		// If verbose, then always print state for json/xml.
+		if (!(dev->dd_flags & DF_ENABLED))
+			xo_emit("{e:state/disabled}");
+		else if (dev->dd_flags & DF_SUSPENDED)
+			xo_emit("{e:state/suspended}");
+		else
+			xo_emit("{e:state/enabled}");
+	}
 
-	printf("%s", dev->dd_name[0] ? dev->dd_name : "unknown");
-	if (vflag && *dev->dd_pnpinfo)
-		printf(" pnpinfo %s", dev->dd_pnpinfo);
-	if (vflag && *dev->dd_location)
-		printf(" at %s", dev->dd_location);
 	if (!(dev->dd_flags & DF_ENABLED))
-		printf(" (disabled)");
+		xo_emit("{D: (disabled)}");
 	else if (dev->dd_flags & DF_SUSPENDED)
-		printf(" (suspended)");
+		xo_emit("{D: (suspended)}");
 }
 
-
 /*
  * Print information about a device.
  */
 int
 print_device(struct devinfo_dev *dev, void *arg)
 {
 	struct indent_arg	ia;
-	int			i, indent;
+	int	indent;
+
+	const char* devname = dev->dd_name[0] ? dev->dd_name : "unknown";
+	int printit = (vflag || (dev->dd_name[0] != 0
+		&& dev->dd_state >= DS_ATTACHED));
 
-	if (vflag || (dev->dd_name[0] != 0 && dev->dd_state >= DS_ATTACHED)) {
+	if (printit) {
```

Comment:
But...  this is used only once, do we need a variable for it?

Reply:
It's used again at the bottom, to close the container.